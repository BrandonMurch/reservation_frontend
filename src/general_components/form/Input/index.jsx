import React, { useState } from 'react';
import PropTypes from 'prop-types';
import style from '../form.module.css';

const Input = function CreateInputAndLabel(props) {
  const [value, setValue] = useState('');
  const [error, setError] = useState('');
  const {
    type,
    name,
    label,
    validator,
  } = props;

  const onBlur = function validateOnBlur({ target }) {
    const errorMessage = validator(target.value);
    let errorFound = false;
    if (errorMessage) {
      setError(errorMessage);
      errorFound = true;
    } else {
      setError('');
    }
    props.onBlur(target.value, target.name, errorFound);
  };

  return (
    <div className={style.inputGroup}>
      <label className={style.labelText} htmlFor={name}>
        {`${label}:`}
      </label>
      <input
        className={style.input}
        value={value}
        onChange={({ target }) => {
          setValue(target.value);
        }}
        id={name}
        onBlur={onBlur}
        type={type}
        name={name}
        required
      />
      {error !== '' && <p className={style.errorText}>{error}</p>}
    </div>
  );
};

Input.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func.isRequired,
  validator: PropTypes.func,
  label: PropTypes.string.isRequired,
};

Input.defaultProps = {
  validator: () => '',
};

export default Input;
