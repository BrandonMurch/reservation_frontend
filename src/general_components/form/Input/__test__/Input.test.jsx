import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';
import renderer from 'react-test-renderer';

import Input from '../index';

describe('<Input />', () => {
  let component;
  let mockValidatorFunction;
  let mockBlurFunction;
  let container = null;
  beforeEach(() => {
    mockBlurFunction = jest.fn();
    mockValidatorFunction = jest.fn(() => 'Error!');
    component = render(<Input
      onBlur={mockBlurFunction}
      type="text"
      name="textInput"
      label="This is a label"
      validator={mockValidatorFunction}
    />);
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  it('should match snapshot', () => {
    const tree = renderer.create(<Input
      onBlur={mockBlurFunction}
      type="text"
      name="textInput"
      label="This is a label"
      validator={mockValidatorFunction}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should have a label and display text', () => {
    const label = component.getByText(/This is a label/i);
    expect(label).toBeInTheDocument();
    // console.log(label);
    expect(label.htmlFor).toEqual('textInput');
  });
  it('should have an input box for text', () => {
    const input = component.getByRole('textbox');
    expect(input.type).toEqual('text');
    expect(input.name).toEqual('textInput');
  });
  it('should validate on blur', () => {
    const input = component.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'input text' } });
    fireEvent.blur(input);
    expect(mockValidatorFunction).toHaveBeenCalledWith('input text');
    expect(mockBlurFunction).toHaveBeenCalledWith('input text', 'textInput', true);
  });
});
