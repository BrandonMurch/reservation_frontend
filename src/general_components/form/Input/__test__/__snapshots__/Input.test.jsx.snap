// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should match snapshot 1`] = `
<div
  className="inputGroup"
>
  <label
    className="labelText"
    htmlFor="textInput"
  >
    This is a label:
  </label>
  <input
    className="input"
    id="textInput"
    name="textInput"
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type="text"
    value=""
  />
</div>
`;
